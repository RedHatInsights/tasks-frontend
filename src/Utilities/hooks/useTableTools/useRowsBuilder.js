import { emptyRows } from '../../../PresentationalComponents/NoResultsTable/NoResultsTable';
import { hasDetails } from '../../../SmartComponents/completedTaskDetailsHelpers';

const columnProp = (column) =>
  column.key ||
  column.original?.toLowerCase() ||
  column.sortByProp?.toLowerCase();

const itemRow = (item, columns, index) => ({
  ...item.rowProps,
  itemId: item.system,
  has_stdout: item.has_stdout,
  display_name: item.display_name,
  jobId: item.id,
  cells: columns.map((column) => ({
    title: column.renderFunc
      ? column.renderFunc(undefined, undefined, item, index)
      : item[columnProp(column)],
  })),
});

const buildRow = (item, columns, rowTransformer, index, parentIndex) => {
  return rowTransformer.flatMap((transformer) => {
    const row = itemRow(item, columns, index);
    return transformer
      ? transformer(row, item, columns, index, parentIndex)
      : row;
  });
};

const useRowsBuilder = (items, columns, options = {}) => {
  const { rowTransformer = [] } = options;
  const EmptyRowsComponent = options.emptyRows || emptyRows;

  const filteredItems = options?.filter ? options.filter(items) : items;

  const sortedItems = options?.sorter
    ? options.sorter(filteredItems)
    : filteredItems;

  const paginatedItems =
    options?.paginator && !options?.isTableLoading
      ? options?.paginator(filteredItems)
      : sortedItems;

  let parentIndex = -1;
  let row;

  const rows =
    paginatedItems.length > 0
      ? paginatedItems.flatMap((item, index) => {
          parentIndex += 1;
          row = buildRow(item, columns, rowTransformer, index, parentIndex);
          if (hasDetails(item)) {
            parentIndex += 1;
          }
          return row;
        })
      : EmptyRowsComponent;

  const pagination = options?.pagination
    ? {
        ...options.pagination,
        itemCount: filteredItems.length,
      }
    : undefined;

  return {
    tableProps: { rows },
    toolbarProps: {
      pagination,
    },
  };
};

export default useRowsBuilder;
