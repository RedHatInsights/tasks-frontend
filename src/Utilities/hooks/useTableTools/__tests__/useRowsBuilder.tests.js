import { renderHook } from '@testing-library/react-hooks';
import items from './__fixtures__/items.fixtures';
import { taskColumnsFixtures as columns } from './__fixtures__/columns.fixtures';
import useRowsBuilder from '../useRowsBuilder';

describe('useRowsBuilder', () => {
  it('returns a rows configuration', () => {
    const paginator = jest.fn((items) => items);

    const { result } = renderHook(() =>
      useRowsBuilder(items, columns, {
        rowTransformer: [undefined],
        isTableLoading: false,
        paginator: paginator,
      })
    );
    expect(paginator).toHaveBeenCalled();
    expect(result).toMatchSnapshot();
  });

  it('doesnt paginate when the table is loading', () => {
    const paginator = jest.fn();
    renderHook(() =>
      useRowsBuilder(items, columns, {
        rowTransformer: [undefined],
        isTableLoading: true,
        paginator: paginator,
      })
    );
    expect(paginator).not.toHaveBeenCalled();
  });
});
