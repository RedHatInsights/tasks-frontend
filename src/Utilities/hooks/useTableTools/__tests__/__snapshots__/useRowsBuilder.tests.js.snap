// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useRowsBuilder returns a rows configuration 1`] = `
Object {
  "all": Array [
    Object {
      "tableProps": Object {
        "rows": Array [
          Object {
            "cells": Array [
              Object {
                "title": "taskA",
              },
              Object {
                "title": 10,
              },
              Object {
                "title": "2022-04-20T10:10:00",
              },
            ],
            "display_name": undefined,
            "has_stdout": undefined,
            "itemId": undefined,
            "jobId": 1,
          },
          Object {
            "cells": Array [
              Object {
                "title": "taskB",
              },
              Object {
                "title": 5,
              },
              Object {
                "title": "2022-04-21T10:10:00",
              },
            ],
            "display_name": undefined,
            "has_stdout": undefined,
            "itemId": undefined,
            "jobId": 2,
          },
          Object {
            "cells": Array [
              Object {
                "title": "taskC",
              },
              Object {
                "title": 3,
              },
              Object {
                "title": "2022-04-25T10:10:00",
              },
            ],
            "display_name": undefined,
            "has_stdout": undefined,
            "itemId": undefined,
            "jobId": 3,
          },
        ],
      },
      "toolbarProps": Object {
        "pagination": undefined,
      },
    },
  ],
  "current": Object {
    "tableProps": Object {
      "rows": Array [
        Object {
          "cells": Array [
            Object {
              "title": "taskA",
            },
            Object {
              "title": 10,
            },
            Object {
              "title": "2022-04-20T10:10:00",
            },
          ],
          "display_name": undefined,
          "has_stdout": undefined,
          "itemId": undefined,
          "jobId": 1,
        },
        Object {
          "cells": Array [
            Object {
              "title": "taskB",
            },
            Object {
              "title": 5,
            },
            Object {
              "title": "2022-04-21T10:10:00",
            },
          ],
          "display_name": undefined,
          "has_stdout": undefined,
          "itemId": undefined,
          "jobId": 2,
        },
        Object {
          "cells": Array [
            Object {
              "title": "taskC",
            },
            Object {
              "title": 3,
            },
            Object {
              "title": "2022-04-25T10:10:00",
            },
          ],
          "display_name": undefined,
          "has_stdout": undefined,
          "itemId": undefined,
          "jobId": 3,
        },
      ],
    },
    "toolbarProps": Object {
      "pagination": undefined,
    },
  },
  "error": undefined,
}
`;
